### PR TITLE
8059632: Method reference compilation uses incorrect qualifying type

### PR DESCRIPTION
**Pull request by sadayapalam**
Align with JLS 13.1 (the qualifying class or interface of the method invocation) by
    referring to the type of the receiver rather than the type of the declaring 
    class/interface in the bootstrap attribute

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8059632](https://bugs.openjdk.org/browse/JDK-8059632): Method reference compilation uses incorrect qualifying type
 * [JDK-8295975](https://bugs.openjdk.org/browse/JDK-8295975): Method reference compilation uses incorrect qualifying type (**CSR**)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [6f39d848](https://git.openjdk.org/jdk/pull/10809/files/6f39d8486b1e132ea5cba6ce0fc47c40ee97ce7e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10809/head:pull/10809` \
`$ git checkout pull/10809`

Update a local copy of the PR: \
`$ git checkout pull/10809` \
`$ git pull https://git.openjdk.org/jdk pull/10809/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10809`

View PR using the GUI difftool: \
`$ git pr show -t 10809`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10809.diff">https://git.openjdk.org/jdk/pull/10809.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/lambda/methodReference/8059632/MethodRefQualifyingTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    public static void main(String argv[]) throws Exception {
+
+        // Without the fix for JDK-8059632, the invocation below will fail with
+        // java.lang.invoke.LambdaConversionException: Invalid receiver type class MethodRefQualifyingTypeTest$MethodSupplierImpl; not a subtype of implementation type interface MethodRefQualifyingTypeTest$MethodSupplier
+
+        // With the fix for JDK-8059632, the invocation would succeed since the bootstrap section
+        // would refer to the type of the receiver and not the type of the declaring interface,
+        // per JLS 13.1 (see "the qualifying type of the method invocation").
+
+        Class.forName("MethodRefQualifyingTypeTest$MethodInvoker").getMethod("invoke").invoke(null);
```
Comment (sadayapalam) [inline test/langtools/tools/javac/lambda/methodReference/8059632/MethodRefQualifyingTypeTest.java:75]: This test is actually derived from a "failing" JCK test: lang/LMBD/lmbd171/lmbd17101m51/lmbd17101m51.html.lmbd17101m51 that incorrectly asserts that there be a linkage error

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 2]
```diff
@@ -3617,6 +3618,56 @@ public int rank(Type t) {
...
+     */
+    public Symbol binaryQualifier(Symbol sym, Type site) {
+
+        if (site.hasTag(ARRAY)) {
+            if (sym == syms.lengthVar ||
+                    sym.owner != syms.arrayClass)
+                return sym;
+            // array clone can be qualified by the array type in later targets
+            Symbol qualifier;
+            if ((qualifier = qualifiedSymbolCache.get(site)) == null) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:3651]: I know this code has been moved here from Gen - I'd be very very curious to understand if the caching mechanism actually does anything meaningful performance-wise. Seems just a way to add more state.

**Review by mcimadamore — APPROVED** [review]
Looks very good to me. Nice unification of the code, the spec is quite clear on this (binary form of method references and method invocation should be the same).

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java (hunk 2 above)
Comment (mcimadamore) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:3651]: Needs a CSR

File: test/langtools/tools/javac/lambda/methodReference/8059632/MethodRefQualifyingTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    interface MyFunctionalInterface {
+        int invokeMethodReference(int a);
+    }
+
+    static class MethodInvoker {
+        public static void invoke() {
+            MyFunctionalInterface instance = null;
+            MethodSupplierImpl ms = new MethodSupplierImpl() {
+                    public  int m(int a){
```
Comment (turbanoff) [inline test/langtools/tools/javac/lambda/methodReference/8059632/MethodRefQualifyingTypeTest.java:57]:
```suggestion
                    public int m(int a){
```

**Pull request closed without merging**